plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinConverter.kt

```diff
@@ -0,0 +1,210 @@
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.asJava.LightClassUtil
+import org.jetbrains.kotlin.asJava.classes.KtLightClass
+import org.jetbrains.kotlin.asJava.elements.*
+import org.jetbrains.kotlin.asJava.toLightClass
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.uast.*
+
+internal object FirKotlinConverter {
+    internal fun convertDeclarationOrElement(
+        element: PsiElement,
+        givenParent: UElement?,
+        requiredTypes: Array<out Class<out UElement>>
+    ): UElement? {
+        if (element is UElement) return element
+
+        // TODO: cache inside PsiElement?
+
+        return convertDeclaration(element, givenParent, requiredTypes)
+            ?: convertPsiElement(element, givenParent, requiredTypes)
+    }
+
+    internal fun convertDeclaration(
+        element: PsiElement,
+        givenParent: UElement?,
+        requiredTypes: Array<out Class<out UElement>>
+    ): UElement? {
+        val original = element.originalElement
+
+        fun <P : PsiElement> build(ctor: (P, UElement?) -> UElement): () -> UElement? = {
+            @Suppress("UNCHECKED_CAST")
+            ctor(original as P, givenParent)
+        }
+
+        fun <P : PsiElement, K : KtElement> buildKt(ktElement: K, ctor: (P, K, UElement?) -> UElement): () -> UElement? = {
+            @Suppress("UNCHECKED_CAST")
+            ctor(original as P, ktElement, givenParent)
+        }
+
+        fun <P : PsiElement, K : KtElement> buildKtOpt(ktElement: K?, ctor: (P, K?, UElement?) -> UElement): () -> UElement? = {
+            @Suppress("UNCHECKED_CAST")
+            ctor(original as P, ktElement, givenParent)
+        }
+
+        return with(requiredTypes) {
+            when (original) {
+                is KtFile -> {
+                    convertKtFile(original, givenParent, requiredTypes).firstOrNull()
+                }
+                is FakeFileForLightClass -> {
+                    el<UFile> { FirKotlinUFile(original.navigationElement) }
+                }
+
+                is KtClassOrObject -> {
+                    convertClassOrObject(original, givenParent, requiredTypes).firstOrNull()
+                }
+                is KtLightClass -> {
+                    // TODO: differentiate enum entry
+                    el<UClass> { AbstractFirKotlinUClass.create(original, givenParent) }
+                }
+                // TODO: KtEnumEntry
+
+                is KtLightField -> {
+                    // TODO: differentiate enum constant
+                    el<UField>(buildKtOpt(original.kotlinOrigin, ::FirKotlinUField))
+                }
+                is KtLightParameter -> {
+                    el<UParameter>(buildKtOpt(original.kotlinOrigin, ::FirKotlinUParameter))
+                }
+                // TODO: UastKotlinPsiParameter ?
+                // TODO: UastKotlinPsiParameterBase ?
+                // TODO: UastKotlinPsiVariable ?
```

Comment:
@xiexed These `UastKotlin*` declarations look quite specific to `uast-kotlin`. Could you explain what those are, and when we will likely see those?

Reply:
They are synthetic elements which are produced at some kind of lowering done by UAST: convertig elvis operator to if, ot local functions to lambdas. Whrn passed to converter these elements should be converted to corresponding Uast elements, actually they could be quite ethemeral, I mean they could have no representation in PSI or FIR

Reply:
Got it! So, we'll see wether we'll need similar lowering in FIR UAST or those are already handled well in FIR/KtSymbols. Hope the latter is the case.